File: plugin-rust-server/src/main/java/jetbrains/buildServer/rust/CargoParametersProvider.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.rust;

import jetbrains.buildServer.rust.commands.CommandType;
import jetbrains.buildServer.rust.commands.cargo.*;

import java.util.Arrays;
import java.util.List;

/**
 * Provides parameters for cargo runner.
 */
public class CargoParametersProvider {

    private final List<CommandType> myTypes;

    public CargoParametersProvider() {
        myTypes = Arrays.asList(
                new BenchCommandType(),
                new BuildCommandType(),
                new CleanCommandType(),
                new DocCommandType(),
                new LoginCommandType(),
                new PackageCommandType(),
                new PublishCommandType(),
                new RunCommandType(),
                new RustcCommandType(),
                new RustDocCommandType(),
                new TestCommandType(),
                new UpdateCommandType(),
                new YankCommandType());
    }

    public List<CommandType> getTypes() {
        return myTypes;
    }

    public String getCommandKey() {
        return CargoConstants.PARAM_COMMAND;
    }

    public String getVerbosityKey() {
        return CargoConstants.PARAM_VERBOSITY;
    }

    public String getBuildPackageKey() {
        return CargoConstants.PARAM_BUILD_PACKAGE;
    }

    public String getBuildTypeKey() {
        return CargoConstants.PARAM_BUILD_TYPE;
    }

    public String getBuildTypeNameKey() {
        return CargoConstants.PARAM_BUILD_TYPE_NAME;
    }

    public String getBuildFeaturesKey() {
        return CargoConstants.PARAM_BUILD_FEATURES;
    }

    public String getBuildNoDefaultFeaturesKey() {
        return CargoConstants.PARAM_BUILD_NO_DEFAULT_FEATURES;
    }

    public String getBuildReleaseKey() {
        return CargoConstants.PARAM_BUILD_RELEASE;
    }

    public String getBuildTargetKey() {
        return CargoConstants.PARAM_BUILD_TARGET;
    }

    public String getBuildManifestKey() {
        return CargoConstants.PARAM_BUILD_MANIFEST;
    }

    public String getBuildParallelKey() {
        return CargoConstants.PARAM_BUILD_PARALLEL;
    }

    public String getCleanPackageKey() {
        return CargoConstants.PARAM_CLEAN_PACKAGE;
    }

    public String getCleanReleaseKey() {
        return CargoConstants.PARAM_CLEAN_RELEASE;
    }

    public String getCleanTargetKey() {
        return CargoConstants.PARAM_CLEAN_TARGET;
    }

    public String getCleanManifestKey() {
        return CargoConstants.PARAM_CLEAN_MANIFEST;
    }

    public String getTestArgumentsKey() {
        return CargoConstants.PARAM_TEST_ARGUMENTS;
    }

    public String getTestPackageKey() {
        return CargoConstants.PARAM_TEST_PACKAGE;
    }

    public String getTestTypeKey() {
        return CargoConstants.PARAM_TEST_TYPE;
    }

    public String getTestTypeNameKey() {
        return CargoConstants.PARAM_TEST_TYPE_NAME;
    }

    public String getTestReleaseKey() {
        return CargoConstants.PARAM_TEST_RELEASE;
    }

    public String getTestNoRunKey() {
        return CargoConstants.PARAM_TEST_NO_RUN;
    }

    public String getTestFeaturesKey() {
        return CargoConstants.PARAM_TEST_FEATURES;
    }

    public String getTestNoDefaultFeaturesKey() {
        return CargoConstants.PARAM_TEST_NO_DEFAULT_FEATURES;
    }

    public String getTestNoFailFastKey() {
        return CargoConstants.PARAM_TEST_NO_FAIL_FAST;
    }

    public String getTestTargetKey() {
        return CargoConstants.PARAM_TEST_TARGET;
    }

    public String getTestManifestKey() {
        return CargoConstants.PARAM_TEST_MANIFEST;
    }

    public String getTestParallelKey() {
        return CargoConstants.PARAM_TEST_PARALLEL;
    }

    public String getRunArgumentsKey() {
        return CargoConstants.PARAM_RUN_ARGUMENTS;
    }

    public String getRunTypeKey() {
        return CargoConstants.PARAM_RUN_TYPE;
    }

    public String getRunTypeNameKey() {
        return CargoConstants.PARAM_RUN_TYPE_NAME;
    }

    public String getRunReleaseKey() {
        return CargoConstants.PARAM_RUN_RELEASE;
    }

    public String getRunFeaturesKey() {
        return CargoConstants.PARAM_RUN_FEATURES;
    }

    public String getRunNoDefaultFeaturesKey() {
        return CargoConstants.PARAM_RUN_NO_DEFAULT_FEATURES;
    }

    public String getRunTargetKey() {
        return CargoConstants.PARAM_RUN_TARGET;
    }

    public String getRunManifestKey() {
        return CargoConstants.PARAM_RUN_MANIFEST;
    }

    public String getRunParallelKey() {
        return CargoConstants.PARAM_RUN_PARALLEL;
    }

    public String getBenchArgumentsKey() {
        return CargoConstants.PARAM_BENCH_ARGUMENTS;
    }

    public String getBenchPackageKey() {
        return CargoConstants.PARAM_BENCH_PACKAGE;
    }

    public String getBenchTypeKey() {
        return CargoConstants.PARAM_BENCH_TYPE;
    }

    public String getBenchTypeNameKey() {
        return CargoConstants.PARAM_BENCH_TYPE_NAME;
    }

    public String getBenchReleaseKey() {
        return CargoConstants.PARAM_BENCH_RELEASE;
    }

    public String getBenchNoRunKey() {
        return CargoConstants.PARAM_BENCH_NO_RUN;
    }

    public String getBenchFeaturesKey() {
        return CargoConstants.PARAM_BENCH_FEATURES;
    }

    public String getBenchNoDefaultFeaturesKey() {
        return CargoConstants.PARAM_BENCH_NO_DEFAULT_FEATURES;
    }

    public String getBenchTargetKey() {
        return CargoConstants.PARAM_BENCH_TARGET;
    }

    public String getBenchManifestKey() {
        return CargoConstants.PARAM_BENCH_MANIFEST;
    }

    public String getBenchParallelKey() {
        return CargoConstants.PARAM_BENCH_PARALLEL;
    }

    public String getDocPackageKey() {
        return CargoConstants.PARAM_DOC_PACKAGE;
    }

    public String getDocReleaseKey() {
        return CargoConstants.PARAM_DOC_RELEASE;
    }

    public String getDocNoDependenciesKey() {
        return CargoConstants.PARAM_DOC_NO_DEPS;
    }

    public String getDocFeaturesKey() {
        return CargoConstants.PARAM_DOC_FEATURES;
    }

    public String getDocNoDefaultFeaturesKey() {
        return CargoConstants.PARAM_DOC_NO_DEFAULT_FEATURES;
    }

    public String getDocTargetKey() {
        return CargoConstants.PARAM_DOC_TARGET;
    }

    public String getDocManifestKey() {
        return CargoConstants.PARAM_DOC_MANIFEST;
    }

    public String getDocParallelKey() {
        return CargoConstants.PARAM_DOC_PARALLEL;
    }

    public String getPackageNoVerifyKey() {
        return CargoConstants.PARAM_PACKAGE_NO_VERIFY;
    }

    public String getPackageNoMetadataKey() {
        return CargoConstants.PARAM_PACKAGE_NO_METADATA;
    }

    public String getPackageManifestKey() {
        return CargoConstants.PARAM_PACKAGE_MANIFEST;
    }

    public String getPublishHostKey() {
        return CargoConstants.PARAM_PUBLISH_HOST;
    }

    public String getPublishTokenKey() {
        return CargoConstants.PARAM_PUBLISH_TOKEN;
    }

    public String getPublishNoVerifyKey() {
        return CargoConstants.PARAM_PUBLISH_NO_VERIFY;
    }

    public String getPublishManifestKey() {
        return CargoConstants.PARAM_PUBLISH_MANIFEST;
    }

    public String getRustcOptsKey() {
        return CargoConstants.PARAM_RUSTC_OPTS;
    }

    public String getRustcPackageKey() {
        return CargoConstants.PARAM_RUSTC_PACKAGE;
    }

    public String getRustcTypeKey() {
        return CargoConstants.PARAM_RUSTC_TYPE;
    }

    public String getRustcTypeNameKey() {
        return CargoConstants.PARAM_RUSTC_TYPE_NAME;
    }

    public String getRustcReleaseKey() {
        return CargoConstants.PARAM_RUSTC_RELEASE;
    }

    public String getRustcFeaturesKey() {
        return CargoConstants.PARAM_RUSTC_FEATURES;
    }

    public String getRustcNoDefaultFeaturesKey() {
        return CargoConstants.PARAM_RUSTC_NO_DEFAULT_FEATURES;
    }

    public String getRustcTargetKey() {
        return CargoConstants.PARAM_RUSTC_TARGET;
    }

    public String getRustcManifestKey() {
        return CargoConstants.PARAM_RUSTC_MANIFEST;
    }

    public String getRustcParallelKey() {
        return CargoConstants.PARAM_RUSTC_PARALLEL;
    }

    public String getLoginTokenKey() {
        return CargoConstants.PARAM_LOGIN_TOKEN;
    }

    public String getLoginHostKey() {
        return CargoConstants.PARAM_LOGIN_HOST;
    }

    public String getUpdatePackageKey() {
        return CargoConstants.PARAM_UPDATE_PACKAGE;
    }

    public String getUpdatePreciseKey() {
        return CargoConstants.PARAM_UPDATE_PRECISE;
    }

    public String getUpdateAggressiveKey() {
        return CargoConstants.PARAM_UPDATE_AGGRESSIVE;
    }

    public String getUpdateManifestKey() {
        return CargoConstants.PARAM_UPDATE_MANIFEST;
    }

    public String getRustDocOptionsKey() {
        return CargoConstants.PARAM_RUSTDOC_OPTS;
    }

    public String getRustDocPackageKey() {
        return CargoConstants.PARAM_RUSTDOC_PACKAGE;
    }

    public String getRustDocTypeKey() {
        return CargoConstants.PARAM_RUSTDOC_TYPE;
    }

    public String getRustDocTypeNameKey() {
        return CargoConstants.PARAM_RUSTDOC_TYPE_NAME;
    }

    public String getRustDocReleaseKey() {
        return CargoConstants.PARAM_RUSTDOC_RELEASE;
    }

    public String getRustDocFeaturesKey() {
        return CargoConstants.PARAM_RUSTDOC_FEATURES;
    }

    public String getRustDocNoDefaultFeaturesKey() {
        return CargoConstants.PARAM_RUSTDOC_NO_DEFAULT_FEATURES;
    }

    public String getRustDocTargetKey() {
        return CargoConstants.PARAM_RUSTDOC_TARGET;
    }

    public String getRustDocManifestKey() {
        return CargoConstants.PARAM_RUSTDOC_MANIFEST;
    }

    public String getRustDocParallelKey() {
        return CargoConstants.PARAM_RUSTDOC_PARALLEL;
    }

    public String getYankCrateKey() {
        return CargoConstants.PARAM_YANK_CRATE;
    }

    public String getYankVersionKey() {
        return CargoConstants.PARAM_YANK_VERSION;
    }

    public String getYankUndoKey() {
        return CargoConstants.PARAM_YANK_UNDO;
    }

    public String getYankIndexKey() {
        return CargoConstants.PARAM_YANK_INDEX;
    }

    public String getYankTokenKey() {
        return CargoConstants.PARAM_YANK_TOKEN;
    }
}

File: plugin-rust-server/src/main/kotlin/jetbrains/buildServer/rust/CargoParametersProvider.kt
/*
 * Copyright 2000-2021 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.rust

import jetbrains.buildServer.controllers.BasePropertiesBean
import jetbrains.buildServer.rust.commands.CommandType
import jetbrains.buildServer.rust.commands.cargo.*
import jetbrains.buildServer.util.StringUtil

/**
 * Provides parameters for cargo runner.
 */
class CargoParametersProvider {

    val types: List<CommandType> = listOf(
            BenchCommandType(),
            BuildCommandType(),
            CleanCommandType(),
            DocCommandType(),
            LoginCommandType(),
            PackageCommandType(),
            PublishCommandType(),
            RunCommandType(),
            RustcCommandType(),
            RustDocCommandType(),
            TestCommandType(),
            UpdateCommandType(),
            YankCommandType()
    )

    val commandKey: String
        get() = CargoConstants.PARAM_COMMAND

    val verbosityKey: String
        get() = CargoConstants.PARAM_VERBOSITY

    val toolchainKey: String
        get() = CargoConstants.PARAM_TOOLCHAIN

    val buildPackageKey: String
        get() = CargoConstants.PARAM_BUILD_PACKAGE

    val buildTypeKey: String
        get() = CargoConstants.PARAM_BUILD_TYPE

    val buildTypeNameKey: String
        get() = CargoConstants.PARAM_BUILD_TYPE_NAME

    val buildFeaturesKey: String
        get() = CargoConstants.PARAM_BUILD_FEATURES

    val buildNoDefaultFeaturesKey: String
        get() = CargoConstants.PARAM_BUILD_NO_DEFAULT_FEATURES

    val buildReleaseKey: String
        get() = CargoConstants.PARAM_BUILD_RELEASE

    val buildTargetKey: String
        get() = CargoConstants.PARAM_BUILD_TARGET

    val buildManifestKey: String
        get() = CargoConstants.PARAM_BUILD_MANIFEST

    val buildParallelKey: String
        get() = CargoConstants.PARAM_BUILD_PARALLEL

    val cleanPackageKey: String
        get() = CargoConstants.PARAM_CLEAN_PACKAGE

    val cleanReleaseKey: String
        get() = CargoConstants.PARAM_CLEAN_RELEASE

    val cleanTargetKey: String
        get() = CargoConstants.PARAM_CLEAN_TARGET

    val cleanManifestKey: String
        get() = CargoConstants.PARAM_CLEAN_MANIFEST

    val testArgumentsKey: String
        get() = CargoConstants.PARAM_TEST_ARGUMENTS

    val testPackageKey: String
        get() = CargoConstants.PARAM_TEST_PACKAGE

    val testTypeKey: String
        get() = CargoConstants.PARAM_TEST_TYPE

    val testTypeNameKey: String
        get() = CargoConstants.PARAM_TEST_TYPE_NAME

    val testReleaseKey: String
        get() = CargoConstants.PARAM_TEST_RELEASE

    val testNoRunKey: String
        get() = CargoConstants.PARAM_TEST_NO_RUN

    val testFeaturesKey: String
        get() = CargoConstants.PARAM_TEST_FEATURES

    val testNoDefaultFeaturesKey: String
        get() = CargoConstants.PARAM_TEST_NO_DEFAULT_FEATURES

    val testNoFailFastKey: String
        get() = CargoConstants.PARAM_TEST_NO_FAIL_FAST

    val testTargetKey: String
        get() = CargoConstants.PARAM_TEST_TARGET

    val testManifestKey: String
        get() = CargoConstants.PARAM_TEST_MANIFEST

    val testParallelKey: String
        get() = CargoConstants.PARAM_TEST_PARALLEL

    val runArgumentsKey: String
        get() = CargoConstants.PARAM_RUN_ARGUMENTS

    val runTypeKey: String
        get() = CargoConstants.PARAM_RUN_TYPE

    val runTypeNameKey: String
        get() = CargoConstants.PARAM_RUN_TYPE_NAME

    val runReleaseKey: String
        get() = CargoConstants.PARAM_RUN_RELEASE

    val runFeaturesKey: String
        get() = CargoConstants.PARAM_RUN_FEATURES

    val runNoDefaultFeaturesKey: String
        get() = CargoConstants.PARAM_RUN_NO_DEFAULT_FEATURES

    val runTargetKey: String
        get() = CargoConstants.PARAM_RUN_TARGET

    val runManifestKey: String
        get() = CargoConstants.PARAM_RUN_MANIFEST

    val runParallelKey: String
        get() = CargoConstants.PARAM_RUN_PARALLEL

    val benchArgumentsKey: String
        get() = CargoConstants.PARAM_BENCH_ARGUMENTS

    val benchPackageKey: String
        get() = CargoConstants.PARAM_BENCH_PACKAGE

    val benchTypeKey: String
        get() = CargoConstants.PARAM_BENCH_TYPE

    val benchTypeNameKey: String
        get() = CargoConstants.PARAM_BENCH_TYPE_NAME

    val benchReleaseKey: String
        get() = CargoConstants.PARAM_BENCH_RELEASE

    val benchNoRunKey: String
        get() = CargoConstants.PARAM_BENCH_NO_RUN

    val benchFeaturesKey: String
        get() = CargoConstants.PARAM_BENCH_FEATURES

    val benchNoDefaultFeaturesKey: String
        get() = CargoConstants.PARAM_BENCH_NO_DEFAULT_FEATURES

    val benchTargetKey: String
        get() = CargoConstants.PARAM_BENCH_TARGET

    val benchManifestKey: String
        get() = CargoConstants.PARAM_BENCH_MANIFEST

    val benchParallelKey: String
        get() = CargoConstants.PARAM_BENCH_PARALLEL

    val docPackageKey: String
        get() = CargoConstants.PARAM_DOC_PACKAGE

    val docReleaseKey: String
        get() = CargoConstants.PARAM_DOC_RELEASE

    val docNoDependenciesKey: String
        get() = CargoConstants.PARAM_DOC_NO_DEPS

    val docFeaturesKey: String
        get() = CargoConstants.PARAM_DOC_FEATURES

    val docNoDefaultFeaturesKey: String
        get() = CargoConstants.PARAM_DOC_NO_DEFAULT_FEATURES

    val docTargetKey: String
        get() = CargoConstants.PARAM_DOC_TARGET

    val docManifestKey: String
        get() = CargoConstants.PARAM_DOC_MANIFEST

    val docParallelKey: String
        get() = CargoConstants.PARAM_DOC_PARALLEL

    val packageNoVerifyKey: String
        get() = CargoConstants.PARAM_PACKAGE_NO_VERIFY

    val packageNoMetadataKey: String
        get() = CargoConstants.PARAM_PACKAGE_NO_METADATA

    val packageManifestKey: String
        get() = CargoConstants.PARAM_PACKAGE_MANIFEST

    val publishHostKey: String
        get() = CargoConstants.PARAM_PUBLISH_HOST

    val publishTokenKey: String
        get() = CargoConstants.PARAM_PUBLISH_TOKEN

    val publishTokenKeySecure: String
        get() = CargoConstants.PARAM_PUBLISH_TOKEN_SECURE

    val publishNoVerifyKey: String
        get() = CargoConstants.PARAM_PUBLISH_NO_VERIFY

    val publishManifestKey: String
        get() = CargoConstants.PARAM_PUBLISH_MANIFEST

    val rustcOptsKey: String
        get() = CargoConstants.PARAM_RUSTC_OPTS

    val rustcPackageKey: String
        get() = CargoConstants.PARAM_RUSTC_PACKAGE

    val rustcTypeKey: String
        get() = CargoConstants.PARAM_RUSTC_TYPE

    val rustcTypeNameKey: String
        get() = CargoConstants.PARAM_RUSTC_TYPE_NAME

    val rustcReleaseKey: String
        get() = CargoConstants.PARAM_RUSTC_RELEASE

    val rustcFeaturesKey: String
        get() = CargoConstants.PARAM_RUSTC_FEATURES

    val rustcNoDefaultFeaturesKey: String
        get() = CargoConstants.PARAM_RUSTC_NO_DEFAULT_FEATURES

    val rustcTargetKey: String
        get() = CargoConstants.PARAM_RUSTC_TARGET

    val rustcManifestKey: String
        get() = CargoConstants.PARAM_RUSTC_MANIFEST

    val rustcParallelKey: String
        get() = CargoConstants.PARAM_RUSTC_PARALLEL

    val loginTokenKey: String
        get() = CargoConstants.PARAM_LOGIN_TOKEN

    val loginTokenKeySecure: String
        get() = CargoConstants.PARAM_LOGIN_TOKEN_SECURE

    val loginHostKey: String
        get() = CargoConstants.PARAM_LOGIN_HOST

    val updatePackageKey: String
        get() = CargoConstants.PARAM_UPDATE_PACKAGE

    val updatePreciseKey: String
        get() = CargoConstants.PARAM_UPDATE_PRECISE

    val updateAggressiveKey: String
        get() = CargoConstants.PARAM_UPDATE_AGGRESSIVE

    val updateManifestKey: String
        get() = CargoConstants.PARAM_UPDATE_MANIFEST

    val rustDocOptionsKey: String
        get() = CargoConstants.PARAM_RUSTDOC_OPTS

    val rustDocPackageKey: String
        get() = CargoConstants.PARAM_RUSTDOC_PACKAGE

    val rustDocTypeKey: String
        get() = CargoConstants.PARAM_RUSTDOC_TYPE

    val rustDocTypeNameKey: String
        get() = CargoConstants.PARAM_RUSTDOC_TYPE_NAME

    val rustDocReleaseKey: String
        get() = CargoConstants.PARAM_RUSTDOC_RELEASE

    val rustDocFeaturesKey: String
        get() = CargoConstants.PARAM_RUSTDOC_FEATURES

    val rustDocNoDefaultFeaturesKey: String
        get() = CargoConstants.PARAM_RUSTDOC_NO_DEFAULT_FEATURES

    val rustDocTargetKey: String
        get() = CargoConstants.PARAM_RUSTDOC_TARGET

    val rustDocManifestKey: String
        get() = CargoConstants.PARAM_RUSTDOC_MANIFEST

    val rustDocParallelKey: String
        get() = CargoConstants.PARAM_RUSTDOC_PARALLEL

    val yankCrateKey: String
        get() = CargoConstants.PARAM_YANK_CRATE

    val yankVersionKey: String
        get() = CargoConstants.PARAM_YANK_VERSION

    val yankUndoKey: String
        get() = CargoConstants.PARAM_YANK_UNDO

    val yankIndexKey: String
        get() = CargoConstants.PARAM_YANK_INDEX

    val yankTokenKey: String
        get() = CargoConstants.PARAM_YANK_TOKEN

    val yankTokenKeySecure: String
        get() = CargoConstants.PARAM_YANK_TOKEN_SECURE

    companion object {
        @JvmStatic
        fun replaceWithNewParameter(propertiesBean: BasePropertiesBean, oldParameterName: String, newParameterName: String) {
            if (!propertiesBean.properties.containsKey(newParameterName) && propertiesBean.properties.containsKey(oldParameterName)) {
                propertiesBean.properties[newParameterName] = propertiesBean.properties[oldParameterName]
                propertiesBean.properties.remove(oldParameterName)
            }
        }
    }
}